win32ss/user/ntuser/defwnd.c

```diff
@@ -790,112 +790,65 @@ IntDefWindowProc(
-                   {
-                       co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_MINIMIZE, lParam);
-                   }
+                       co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_MINIMIZE, MAKELONG(0, 1));
                }
                else if (wParam == VK_UP)
                {
-                  RECT currentRect;
-                  if ((topWnd->InternalPos.NormalRect.right == topWnd->InternalPos.NormalRect.left) ||
-                      (topWnd->InternalPos.NormalRect.top == topWnd->InternalPos.NormalRect.bottom))
-                  {
-                      currentRect = topWnd->rcWindow;
-                  }
-                  else
-                  {
-                      currentRect = topWnd->InternalPos.NormalRect;
-                  }
-                  co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_MAXIMIZE, 0);
-
-                  // save normal rect if maximazing snapped window
-                  topWnd->InternalPos.NormalRect = currentRect;
+                   if (topWnd->style & WS_MINIMIZE)
+                       co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_RESTORE, MAKELONG(0, 1));
+                   else
+                       co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_MAXIMIZE, MAKELONG(0, 1));
                }
                else if (wParam == VK_LEFT || wParam == VK_RIGHT)
                {
-                  RECT snapRect, normalRect, windowRect;
-                  BOOL snapped;
-                  normalRect = topWnd->InternalPos.NormalRect;
-                  snapped = (normalRect.left != 0 && normalRect.right != 0 &&
-                             normalRect.top != 0 && normalRect.bottom != 0);
+                  UINT edge = wParam == VK_LEFT ? HTLEFT : HTRIGHT;
+                  UINT otherEdge = edge == HTLEFT ? HTRIGHT : HTLEFT;
 
                   if (topWnd->style & WS_MAXIMIZE)
                   {
-                     co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_RESTORE, lParam);
-                     snapped = FALSE;
-                  }
-                  windowRect = topWnd->rcWindow;
-
-                  UserSystemParametersInfo(SPI_GETWORKAREA, 0, &snapRect, 0);
-                  if (wParam == VK_LEFT)
-                  {
-                     snapRect.right = (snapRect.left + snapRect.right) / 2;
+                     /* SC_RESTORE + Snap causes the window to visually move twice, place it manually in the snap position */
+                     RECT normalRect = topWnd->InternalPos.NormalRect;
+                     co_IntCalculateSnapPosition(topWnd, edge, &topWnd->InternalPos.NormalRect); /* Place us here please */
+                     IntSetSnapEdge(topWnd, edge); /* Tell everyone the edge we are snapped to */
+                     co_IntSendMessage(hwndTop, WM_SYSCOMMAND, SC_RESTORE, MAKELONG(0, 1));
+                     IntSetSnapInfo(topWnd, edge, &normalRect); /* Reset the real place to unsnap to */
+                     snapped = FALSE; /* Force snap */
```

Comment:
The ill-named "snapped" is not a boolean, so why do you use "FALSE" here?

Reply:
The snap state is designed to work as a BOOL if you don't care about which edge. I just like the feel of it that way but I can use HTNOWHERE...